src/backend/d3d12/BufferD3D12.cpp

```diff
@@ -113,31 +118,35 @@ namespace backend { namespace d3d12 {
         return mResource;
     }
 
-    bool Buffer::GetResourceTransitionBarrier(nxt::BufferUsageBit currentUsage,
-                                              nxt::BufferUsageBit targetUsage,
-                                              D3D12_RESOURCE_BARRIER* barrier) {
-        if (GetAllowedUsage() & (nxt::BufferUsageBit::MapRead | nxt::BufferUsageBit::MapWrite)) {
-            // Transitions are never needed for mapped buffers because they are created with and
-            // always need the Transfer(Dst|Src) state. Mapped buffers cannot have states outside of
-            // (MapRead|TransferDst) and (MapWrite|TransferSrc)
-            return false;
+    void Buffer::TransitionUsageNow(ComPtr<ID3D12GraphicsCommandList> commandList,
+                                    nxt::BufferUsageBit usage) {
+        // Resources in upload and readback heaps must be kept in the COPY_SOURCE/DEST state
+        if (mFixedResourceState) {
+            ASSERT(usage == mLastUsage);
+            return;
         }
 
-        D3D12_RESOURCE_STATES stateBefore = D3D12BufferUsage(currentUsage);
-        D3D12_RESOURCE_STATES stateAfter = D3D12BufferUsage(targetUsage);
-
-        if (stateBefore == stateAfter) {
-            return false;
+        // We can skip transitions to already current usages.
+        // TODO(cwallez@chromium.org): Need some form of UAV barriers at some point.
+        bool lastIncludesTarget = (mLastUsage & usage) == usage;
+        if (lastIncludesTarget) {
+            return;
         }
 
-        barrier->Type = D3D12_RESOURCE_BARRIER_TYPE_TRANSITION;
-        barrier->Flags = D3D12_RESOURCE_BARRIER_FLAG_NONE;
-        barrier->Transition.pResource = mResource.Get();
-        barrier->Transition.StateBefore = stateBefore;
-        barrier->Transition.StateAfter = stateAfter;
-        barrier->Transition.Subresource = 0;
+        D3D12_RESOURCE_STATES lastState = D3D12BufferUsage(mLastUsage);
```

Comment:
Nit (or personal style, and not new to this patch): it'd be nice if this helper func was ToD3D12BufferUsage(), so it didn't look like a constructor.

Reply:
I wouldn't mind that change but it would be best to do it across all backends at once so I'll leave it for later.